### PR TITLE
Tuning cgroup

### PR DESCRIPTION
**Pull request by aherrmann-suse**
### Description
Describe (blk)io controller when using BFQ and cgroups.
Currently there is no description at all about this.

A somehow related bug report is
[Bug 1170566] L3: IO accounting/scheduler does not work

### Checklist
* Check all items that apply.

*Are backports required?*

- [ ] To maintenance/SLE15SP1
- [ ] To maintenance/SLE15SP0
- [ ] To maintenance/SLE12SP5
- [ ] To maintenance/SLE12SP4
- [ ] To maintenance/SLE12SP3

### CODE REVIEW

**Comment by cjschroder** [comment]
Hello @aherrmann-suse! Thank you for this :) There are some markup corrections that I will make. 

re:

````
blkio.weight (only available in older kernels with legacy block layer and using the CFQ I/O scheduler)
blkio.bfq.weight (only available in newer kernels with blk-mq and using BFQ I/O scheduler) 
````

Can you be a little more specific than 'older' and 'newer' kernels, do you know which kernel versions?

Should this be backported?

thanks!

**Comment by aherrmann-suse** [comment]
On Tue, Jul 14, 2020 at 01:32:03PM -0700, Carla Schroder wrote:
> Hello @aherrmann-suse! Thank you for this :) There are some markup corrections
> that I will make.

Great, thanks.

> re:
> 
> blkio.weight (only available in older kernels with legacy block layer and using the CFQ I/O scheduler)

maybe I should write

(available in kernels up to version 4.20 with legacy block layer when
using CFQ I/O scheduler)

> blkio.bfq.weight (only available in newer kernels with blk-mq and using BFQ I/O scheduler)

(available in kernels starting with version 5.0 (with blk-mq when
using BFQ I/O scheduler)

> Can you be a little more specific than 'older' and 'newer' kernels, do you know
> which kernel versions?

Yes, see above.
(Commit f382fb0bcef4c ("block: remove legacy IO schedulers")
etc. removed the old code in kernel v5.0. Thus older refers to pre
v5.0 kernels and newer to kernels starting with v5.0.)

> Should this be backported?

I wasn't sure. But re-thinking about it:
A backport to maintenance branches for SLE15-SP1 and SLE12-SP5 might
make sense.

Should I prepare a new pull request?

> thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.*
> 

-- 
Regards,
Andreas

SUSE Software Solutions Germany GmbH
Maxfeldstr. 5, 90409 Nürnberg, Germany
GF: Felix Imendörffer (HRB 36809, AG Nürnberg)

File: xml/tuning_cgroups.xml [hunk 1]
```diff
@@ -386,14 +386,181 @@ TasksMax=16284
...
+131072+0 records in
+131072+0 records out
+536870912 bytes (537 MB, 512 MiB) copied, 1.69026 s, 318 MB/s
+</screen>
+  </sect2>
+  <sect2>
+   <title>Using cgroup-v2</title>
+   <para>
+    First we need to make sure that Block IO controller is not active
+    under cgroup-v1. Thus we boot with kernel parameter
```
Comment (dmpop) [inline xml/tuning_cgroups.xml:505]:
```suggestion
    under cgroup-v1. To do this, boot with kernel parameter
```

File: xml/tuning_cgroups.xml [hunk 2]
```diff
@@ -386,14 +386,181 @@ TasksMax=16284
...
+    <option>cgroup_no_v1=blkio</option>. Block IO controller is
+    inactive in cgroup-v1 and IO controller shows up as an option for
```
Comment (dmpop) [inline xml/tuning_cgroups.xml:507]:
```suggestion
    inactive in cgroup-v1, and IO controller shows up as an option for
```

File: xml/tuning_cgroups.xml [hunk 3]
```diff
@@ -386,14 +386,181 @@ TasksMax=16284
...
+<screen>
+$ cat /proc/cmdline
+BOOT_IMAGE=... cgroup_no_v1=blkio ...
+$ ls /sys/fs/cgroup/blkio/
+$ cat /sys/fs/cgroup/unified/cgroup.controllers
+io
+</screen>
+ <para>
+   Next steps are: enable IO controller, create 2 control groups,
+   ensure that BFQ is used, set weights.
```
Comment (dmpop) [inline xml/tuning_cgroups.xml:519]:
```suggestion
   Next, enable IO controller, create 2 control groups,
   ensure that BFQ is used, and set weights:
```

File: xml/tuning_cgroups.xml [hunk 4]
```diff
@@ -386,14 +386,181 @@ TasksMax=16284
...
+[I/O controller] /sys/fs/cgroup/unified> cat /sys/block/sda/queue/scheduler
+mq-deadline kyber [bfq] none
+[I/O controller] /sys/fs/cgroup/unified> cat reader1/io.bfq.weight
+default 100
+[I/O controller] /sys/fs/cgroup/unified> sudo su -c "echo 400 >reader2/io.bfq.weight"
+[I/O controller] /sys/fs/cgroup/unified> cat reader2/io.bfq.weight
+default 400
+</screen>
+   <para>
+    With this setting throughput looks like:
```
Comment (dmpop) [inline xml/tuning_cgroups.xml:539]:
```suggestion
    With these settings, throughput looks like:
```

**Comment by cjschroder** [comment]
> On Tue, Jul 14, 2020 at 01:32:03PM -0700, Carla Schroder wrote: Hello @aherrmann-suse! Thank you for this :) There are some markup corrections that I will make.
> Great, thanks.
> re: blkio.weight (only available in older kernels with legacy block layer and using the CFQ I/O scheduler)
> maybe I should write (available in kernels up to version 4.20 with legacy block layer when using CFQ I/O scheduler)
> blkio.bfq.weight (only available in newer kernels with blk-mq and using BFQ I/O scheduler)
> (available in kernels starting with version 5.0 (with blk-mq when using BFQ I/O scheduler)
> Can you be a little more specific than 'older' and 'newer' kernels, do you know which kernel versions?
> Yes, see above. (Commit f382fb0bcef4c ("block: remove legacy IO schedulers") etc. removed the old code in kernel v5.0. Thus older refers to pre v5.0 kernels and newer to kernels starting with v5.0.)
> Should this be backported?
> I wasn't sure. But re-thinking about it: A backport to maintenance branches for SLE15-SP1 and SLE12-SP5 might make sense. Should I prepare a new pull request?


Yes, please add the lines about kernel versions.

I think we can continue with this PR, what you think @dmpop ? We can edit directly using the Github editing form, how do you feel about incorporating your changes?

**Comment by cjschroder** [comment]
With the release of SLE 15 SP3 tomorrow, the current master branch will be renamed maintenance/SLE15SP2, and master will become SLE 15 SP3. Keep this in mind when it comes time to merge this PR & then do the backports.

**Comment by dmpop** [comment]
> I think we can continue with this PR, what you think @dmpop ? We can edit directly using the Github editing form, how do you feel about incorporating your changes?

Sure, let's just keep working on this PR. Once it's merged, we can backport it ourselves. @aherrmann-suse Please review and accept/reject my suggestions. :-)

**Comment by aherrmann-suse** [comment]
On Tue, Jul 21, 2020 at 02:24:40AM -0700, Dmitri Popov wrote:
>     I think we can continue with this PR, what you think @dmpop ? We can edit
>     directly using the Github editing form, how do you feel about incorporating
>     your changes?
> 
> Sure, let's just keep working on this PR. Once it's merged, we can backport it
> ourselves. @aherrmann-suse Please review and accept/reject my suggestions. :-)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.*

I've accepted your suggestions.
I am unsure how to modify the information about kernel versions in this PR.

-- 
Regards,
Andreas

SUSE Software Solutions Germany GmbH
Maxfeldstr. 5, 90409 Nürnberg, Germany
GF: Felix Imendörffer (HRB 36809, AG Nürnberg)

**Comment by dmpop** [comment]
> I've accepted your suggestions.

Thank you, Andreas!

> I am unsure how to modify the information about kernel versions in this PR.

Make edits in the `tuning_cgroup`  branch of your `doc-sle` repo fork. Push then the changes, and the PR will be automatically updated.

**Comment by aherrmann-suse** [comment]
On Tue, Jul 21, 2020 at 05:04:26AM -0700, Dmitri Popov wrote:
>     I've accepted your suggestions.
> 
> Thank you, Andreas!
> 
>     I am unsure how to modify the information about kernel versions in this PR.
> 
> Make edits in the tuning_cgroup branch of your doc-sle repo fork. Push then the
> changes, and the PR will be automatically updated.

I've done that. Hope it's ok now.

-- 
Regards,
Andreas

SUSE Software Solutions Germany GmbH
Maxfeldstr. 5, 90409 Nürnberg, Germany
GF: Felix Imendörffer (HRB 36809, AG Nürnberg)

**Comment by dmpop** [comment]
> I've done that. Hope it's ok now.

Perfect! Thank you very much. @cjschroder If it look goods to you, I'll merge the PR.

**Comment by cjschroder** [comment]
Sorry friends, we are not done yet!  I have questions! I tried the steps in the cgroups v2 section. (5.3.18-lp152.20.7-default kernel, opensuse leap 15.2) I have not tested this on SLES yet.

I created proper command prompts-- is this how it is supposed to look? We need to present to our readers as exactly as possible what the commands and output are supposed to look like. (see attached PDF)

Where does `/mnt/tests/io-cgroup` come from? 

Where do `file1` and `file` come from? They do not appear to be empty test files, because your example output shows they have something in them.

When I run `sudo su -c "echo '+io' > cgroup.subtree_control"` I get an error:

````
bash: line 0: echo: write error: No such file or directorysudo su -c "echo 5220 > reader1/cgroup.procs"
````
But `cat /sys/fs/cgroup/unified/cgroup.controllers` returns nothing, so maybe I missed some steps? I will try again tomorrow.

Thanks in advance, Andreas!

@aherrmann-suse @dmpop

**Comment by cjschroder** [comment]
allrightythen, today's story: I tested the `9.4.2 Using cgroup-v2` section on SLE 15 SP2. I followed each step exactly. Every step was successful until: 

````
$ sudo su -c "echo 3712 > reader1/cgroup.procs"
bash: line 0: echo: write error: No such process
````

The reader1 and reader2 directories are populated:

````
ls reader1/
cgroup.controllers      cgroup.procs            cpu.pressure   io.pressure
cgroup.events           cgroup.stat             cpu.stat       io.stat
cgroup.freeze           cgroup.subtree_control  io.bfq.weight  io.weight
cgroup.max.depth        cgroup.threads          io.latency     memory.pressure
cgroup.max.descendants  cgroup.type             io.max
````

I tested the `sudo su -c "echo 3712 > reader1/cgroup.procs"` in my homedir just to make sure, and it worked. So, something seems to be missing

**Comment by aherrmann-suse** [comment]
On Tue, Jul 21, 2020 at 05:49:52PM -0700, Carla Schroder wrote:
> Sorry friends, we are not done yet! I have questions! I tried the steps in the
> cgroups v2 section. (5.3.18-lp152.20.7-default kernel, opensuse leap 15.2) I
> have not tested this on SLES yet.
> 
> I created proper command prompts-- is this how it is supposed to look? We need
> to present to our readers as exactly as possible what the commands and output
> are supposed to look like. (see attached PDF)

I've used the command prompts to present additional information:
1. What shell is used. (reader1, reader2, "I/O controller")
2. What is the current working directory.

Replacing the prompts deletes this information and the example doesn't
make sense anymore.

> Where does /mnt/tests/io-cgroup come from?

Not explained, just a test directory where some files file1 and file2
exist.

> Where do file1 and file come from? They do not appear to be empty test files,
> because your example output shows they have something in them.

Not empty. Large enough to do the specified read operations.

> When I run sudo su -c "echo '+io' > cgroup.subtree_control" I get an error:

Have you booted your system with cgroup_no_v1=blkio kernel command
line?

> bash: line 0: echo: write error: No such file or directorysudo su -c "echo 5220 > reader1/cgroup.procs"
>
> But cat /sys/fs/cgroup/unified/cgroup.controllers returns nothing, so maybe I
> missed some steps? I will try again tomorrow.
> 
> Thanks in advance, Andreas!
> 
> @aherrmann-suse @dmpop
> 

I admit the examples are kind of brief. I omitted information that I
thought is superfluous for some experienced reader dealing with
cgroups.


-- 
Regards,
Andreas

SUSE Software Solutions Germany GmbH
Maxfeldstr. 5, 90409 Nürnberg, Germany
GF: Felix Imendörffer (HRB 36809, AG Nürnberg)

**Comment by aherrmann-suse** [comment]
On Wed, Jul 22, 2020 at 01:52:33PM -0700, Carla Schroder wrote:
> allrightythen, today's story: I tested the 9.4.2 Using cgroup-v2 section on SLE
> 15 SP2. I followed each step exactly. Every step was successful until:
> 
> $ sudo su -c "echo 3712 > reader1/cgroup.procs"
> bash: line 0: echo: write error: No such process

The number 3712 is not fixed. It's the process ID (PID) of the shell from
which reader1 is started. Ie. reader1 is doing
> echo $$; dd if=file1 of=/dev/null bs=4k
3712
^^^^ this is where the PID comes from.

To run this example you need 3 shells. One for each reader and one to do
the steps to control the I/O. When adding the PID to reader1/cgroup.procs
it means that this process and all its child processes are controlled by
settings for this control group.

> The reader1 and reader2 directories are populated:
> 
> ls reader1/
> cgroup.controllers      cgroup.procs            cpu.pressure   io.pressure
> cgroup.events           cgroup.stat             cpu.stat       io.stat
> cgroup.freeze           cgroup.subtree_control  io.bfq.weight  io.weight
> cgroup.max.depth        cgroup.threads          io.latency     memory.pressure
> cgroup.max.descendants  cgroup.type             io.max
> 
> I tested the sudo su -c "echo 3712 > reader1/cgroup.procs" in my homedir just
> to make sure, and it worked. So, something seems to be missing


-- 
Regards,
Andreas

SUSE Software Solutions Germany GmbH
Maxfeldstr. 5, 90409 Nürnberg, Germany
GF: Felix Imendörffer (HRB 36809, AG Nürnberg)

**Comment by cjschroder** [comment]
Thanks @aherrmann-suse, the mysteries they are solved! I will give it another try, I appreciate your fast response. We have to explain everything, and ideally outline the steps so clearly that the reader can succeed on the first try. Even experienced admins have gaps in their knowledge, so we must be as clear as possible.

Don't worry about having to create perfect PRs. When you provide the expert knowledge, which is the important bit, we are happy to do the rest.

I spent a fair bit of time studying the kernel docs (for example https://www.kernel.org/doc/html/latest/block/bfq-iosched.html?highlight=bfq), and they tell everything about how it works, but are not very good at teaching admins how to actually configure it. There is not much good documentation on this subject, so I am pleased that we can add this.

**Comment by cjschroder** [comment]
> With the release of SLE 15 SP3 tomorrow, the current master branch will be renamed maintenance/SLE15SP2, and master will become SLE 15 SP3. 

This is wrong. We are not working on SP3 yet. We're still at SLE 15 SP2.

**Comment by aherrmann-suse** [comment]
On Thu, Jul 23, 2020 at 08:43:59AM -0700, Carla Schroder wrote:
> Thanks @aherrmann-suse, the mysteries they are solved! I will give it another
> try, I appreciate your fast response. We have to explain everything, and
> ideally outline the steps so clearly that the reader can succeed on the first
> try. Even experienced admins have gaps in their knowledge, so we must be as
> clear as possible.
> 
> Don't worry about having to create perfect PRs. When you provide the expert
> knowledge, which is the important bit, we are happy to do the rest.
> 
> I spent a fair bit of time studying the kernel docs (for example https://
> www.kernel.org/doc/html/latest/block/bfq-iosched.html?highlight=bfq), and they
> tell everything about how it works, but are not very good at teaching admins
> how to actually configure it. There is not much good documentation on this
> subject, so I am pleased that we can add this.

On balance:
Should this PR be dismissed and I try to rework the example for better
understanding? And I'll send a new PR when done with it(?).

BTW, I've just figured that the reference given for cgroup-v2 needs a
correction.

-- 
Regards,
Andreas

SUSE Software Solutions Germany GmbH
Maxfeldstr. 5, 90409 Nürnberg, Germany
GF: Felix Imendörffer (HRB 36809, AG Nürnberg)

**Comment by cjschroder** [comment]
Hi @aherrmann-suse, let's keep this PR for now. I will commit a new set of changes today. I figured out a way to use your labeling on the command prompts (reader1, reader1 etc) that conforms to how prompts are supposed to be formatted. I also added a few explanations. These are not major changes, just prettying and and a bit of clarifying. Go ahead and make corrections if you wish. 

Is the cgroups v2 section complete, are there more steps? It seems short compared to the cgroups v1 section.

**Comment by cjschroder** [comment]
Allrightythen, I committed my changes and attached a PDF. Many thanks @aherrmann-suse, this turned into a lot of work and your contributions are invaluable. This is a difficult and highly-technical topic, so it is worth the extra work to teach it as well as we can.

If you prefer to create a new PR that is fine with me. I think the custom prompt entities (at the top of the file) add clarity and save a lot of explanations, because they show which shell the commands are run from, and the current working directory. I elected to use the `io-controller` label on the prompts that you did not label rather than explain 'here, open a fourth terminal for these commands', and ran everything as root for simplicity.

I want to test the example commands one more time, and I will wait for any changes you want to make.

Thanks again, your work and expertise are much appreciated.

[cha-tuning-cgroups_color_draft_en.pdf](https://github.com/SUSE/doc-sle/files/5001884/cha-tuning-cgroups_color_draft_en.pdf)

**Comment by aherrmann-suse** [comment]
Hello Carla,

Thanks for your changes. I think the general layout/content is now in
much better shape.
I have some comments. See below.
I won't make a new PR now. I'll be on vacation the next two weeks.
If below points are addressed I think the PR can be merged.

On Thu, Jul 30, 2020 at 07:43:37AM -0700, Carla Schroder wrote:
> Allrightythen, I committed my changes and attached a PDF. Many thanks
> @aherrmann-suse, this turned into a lot of work and your contributions are
> invaluable. This is a difficult and highly-technical topic, so it is worth the
> extra work to teach it as well as we can.
> 
> If you prefer to create a new PR that is fine with me. I think the custom
> prompt entities (at the top of the file) add clarity and save a lot of
> explanations, because they show which shell the commands are run from, and the
> current working directory. I elected to use the io-controller label on the
> prompts that you did not label rather than explain 'here, open a fourth
> terminal for these commands', and ran everything as root for simplicity.
> 
> I want to test the example commands one more time, and I will wait for any
> changes you want to make.
> 
> Thanks again, your work and expertise are much appreciated.
> 
> cha-tuning-cgroups_color_draft_en.pdf

(1) To create the test files, please use "512MiB" such that the test
files are sufficiently large:

  $ yes this is a test file | head -c 512MiB >| file2.txt
  $ ls -lh file2.txt
  -rw-r--r-- 1 reasand users 512M Jul 31 10:02 file2.txt

--------------------
(2) Section 9.4.1 (1st example)

The read value (190 MB/s) in the first example seems quite low. It
usually should be around 500 MB/s for an SSD, e.g.:

  reasand@suselix:~> sudo su -c "sync; echo 3 > /proc/sys/vm/drop_caches"
  [sudo] password for root:
  reasand@suselix:~> echo $$; dd if=file2.txt of=/dev/null bs=4k count=131072
  5001
  131072+0 records in
  131072+0 records out
  536870912 bytes (537 MB, 512 MiB) copied, 1.01985 s, 526 MB/s

That would also match what's shown later that with two readers each
one gets roughly half the maximum throughput.

--------------------
(3) shell prompts needs adaption (3rd example, Section 9.4.1)

The example after

  "Each process gets half of the throughput for I/O operations. Next,
  set up two control groups—one for each process—verify that BFQ is
  used, and set different weights:"

starts with

  cd /sys/fs/cgroup/blkio/

Thus the prompt should show this directory.

--------------------
(4) Section 9.4.2 (BFQ)

BFQ is an I/O scheduler not a controller, thus

  "Next, enable the IO controller, create two control groups, ensure
  that BFQ is used (the activec ontroller is enclosed in square
  brackets), and set weights:"

should change to

  "Next, enable the IO controller, create two control groups, ensure
  that BFQ is used (the active scheduler is enclosed in square
  brackets), and set weights:"

--------------------
(3) shell prompts needs adaption (2nd example, Section 9.4.2)

After

  cd /sys/fs/cgroup/unified/

the prompt should reflect this change.

-- 
Regards,
Andreas

SUSE Software Solutions Germany GmbH
Maxfeldstr. 5, 90409 Nürnberg, Germany
GF: Felix Imendörffer (HRB 36809, AG Nürnberg)

**Comment by aherrmann-suse** [comment]
On Thu, Jul 30, 2020 at 06:00:22AM -0700, Carla Schroder wrote:
> Hi @aherrmann-suse, let's keep this PR for now. I will commit a new set of
> changes today. I figured out a way to use your labeling on the command prompts
> (reader1, reader1 etc) that conforms to how prompts are supposed to be
> formatted. I also added a few explanations. These are not major changes, just
> prettying and and a bit of clarifying. Go ahead and make corrections if you
> wish.
> 
> Is the cgroups v2 section complete, are there more steps? It seems short
> compared to the cgroups v1 section.

I think it is complete. It points out the differences to the v1
interface (differing file names) and contains the additional step to
enable the 'io' controller.

-- 
Regards,
Andreas

SUSE Software Solutions Germany GmbH
Maxfeldstr. 5, 90409 Nürnberg, Germany
GF: Felix Imendörffer (HRB 36809, AG Nürnberg)

**Comment by cjschroder** [comment]
allrightythen, due to needing to rebase & other complications I made a commit to master, dad4164fe4ab68aa1. I will make the backports when I return next Monday. Anyone wanting to make changes please open a new PR. Thank you all and have a good weekend!

**Pull request closed without merging**